File: Animator/Java/src/Lib/EffectFactory.java
package Lib;

import Utils.Point;

public class EffectFactory {
    private static EffectFactory instance;

    public static EffectFactory getInstance() {
        if (null == instance) {
            instance = new EffectFactory();
        }

        return instance;
    }

    private EffectFactory() {
    }

    public <T extends IMovable> IEffect<T> makeMove(T o, Point p) {
        return (IEffect<T>) new Mover(p);
    }

    public <T extends IRotatable> IEffect<T> makeRotate(T o, double angle) {
        return (IEffect<T>) new Rotator(angle);
    }
}

File: Animator/C#/PatternsAnimator/Lib/EffectFactory.cs
using System;
using PatternsAnimator.Utils;

namespace PatternsAnimator.Lib
{
    public sealed class EffectFactory
    {
        private static readonly Lazy<EffectFactory> Lazy = new Lazy<EffectFactory>(() => new EffectFactory());

        public static EffectFactory Instance => Lazy.Value;

        private EffectFactory()
        {
        }

        public IEffect<T> MakeMove<T>(T _, Point p) where T : IMovable
        {
            return (IEffect<T>) new Mover(p);
        }

        public IEffect<T> MakeRotate<T>(T _, double angle) where T : IRotateable
        {
            return (IEffect<T>) new Rotator(angle);
        }
    }
}